test/integration/pubsub_msg_seen_cache_test.go

```diff
@@ -0,0 +1,244 @@
+	bcfg := bootstrap.BootstrapConfigWithPeers([]peer.AddrInfo{bis})
+	if err = producerNode.Bootstrap(bcfg); err != nil {
+		t.Fatal(err)
+	}
+	if err = consumerNode.Bootstrap(bcfg); err != nil {
+		t.Fatal(err)
+	}
+
+	// Set up the consumer subscription
+	const TopicName = "topic"
+	consumerSubscription, err := consumerAPI.PubSub().Subscribe(ctx, TopicName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Utility functions defined inline to include context in closure
+	now := func() string {
+		return time.Now().Format(time.StampMilli)
+	}
+	ctr := 0
+	msgGen := func() string {
+		ctr++
+		return fmt.Sprintf("msg_%d", ctr)
+	}
+	produceMessage := func() string {
+		msgTxt := msgGen()
+		err = producerAPI.PubSub().Publish(ctx, TopicName, []byte(msgTxt))
+		if err != nil {
+			t.Fatal(err)
+		}
+		return msgTxt
+	}
+	consumeMessage := func(msgTxt string, shouldFind bool) {
+		// Set up a separate timed context for receiving messages
+		rxCtx, rxCancel := context.WithTimeout(context.Background(), time.Second)
+		defer rxCancel()
+		msg, err := consumerSubscription.Next(rxCtx)
+		if shouldFind {
+			if err != nil {
+				t.Logf("did not receive [%s] by [%s]", msgTxt, now())
+				t.Fatal(err)
+			}
+			t.Logf("received [%s] at [%s]", string(msg.Data()), now())
+			if !bytes.Equal(msg.Data(), []byte(msgTxt)) {
+				t.Fatalf("consumed data [%s] does not match published data [%s]", string(msg.Data()), msgTxt)
+			}
+		} else {
+			if err == nil {
+				t.Logf("received [%s] at [%s]", string(msg.Data()), now())
+				t.Fail()
+			}
+			t.Logf("did not receive [%s] by [%s]", msgTxt, now())
+		}
+	}
+
+	// Send message 1 with the message ID we're going to duplicate later
+	sendDupMsg = true
+	msgTxt := produceMessage()
+	consumeMessage(msgTxt, true) // should find message
+
+	// Send message 2 with the same message ID as before
+	sendDupMsg = true
+	msgTxt = produceMessage()
+	consumeMessage(msgTxt, false) // should NOT find message
+
+	// Wait for seen cache TTL time to let seen cache entries time out
+	time.Sleep(ttl)
+
+	// Send message 3 with a new message ID
+	//
+	// This extra step is necessary for testing the cache TTL because the PubSub code only garbage collects when a
+	// message ID was not already present in the cache. This means that message 2's cache entry, even though it has
+	// technically timed out, will still cause the message to be considered duplicate. When a message with a different
+	// ID passes through, it will be added to the cache and garbage collection will clean up message 2's entry. This is
+	// another bug in the pubsub/cache implementation that will be fixed once the code is refactored for this issue:
+	// https://github.com/libp2p/go-libp2p-pubsub/issues/502
+	sendDupMsg = false
+	msgTxt = produceMessage()
+	consumeMessage(msgTxt, true) // should find message
+
+	// Send message 4 with the same message ID as before
+	sendDupMsg = true
+	msgTxt = produceMessage()
+	consumeMessage(msgTxt, true) // should find message
```

Comment:
```suggestion
	consumeMessage(msgTxt, true) // should find message again (time since the last read > SeenMessagesTTL, so it looks like a new message).
```